daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -197,42 +232,48 @@ private[speedy] sealed abstract class SBuiltin(val arity: Int) {
       case SAnyContract(tyCon, value) => (tyCon, value)
       case otherwise => unexpectedType(i, "AnyContract", otherwise)
     }
-
-  final protected def checkToken(args: util.ArrayList[SValue], i: Int): Unit =
-    args.get(i) match {
-      case SToken => ()
-      case otherwise => unexpectedType(i, "SToken", otherwise)
-    }
-
 }
 
 private[speedy] sealed abstract class SBuiltinPure(arity: Int) extends SBuiltin(arity) {
 
+  /** Pure builtins do not modify the machine state and do not ask questions of the ledger. As a result, pure builtin
+    * execution is immediate.
+    *
+    * @param args arguments for executing the pure builtin
+    * @return the pure builtin's resulting value (wrapped as a Control value)
+    */
+  private[speedy] def executePure(args: util.ArrayList[SValue]): SValue
+
   override private[speedy] final def execute(
       args: util.ArrayList[SValue],
       machine: Machine,
   ): Control = {
     Control.Value(executePure(args))
   }
-
-  /** Execute the (pure) builtin with 'arity' number of arguments in 'args'.
-    *    Returns the resulting value
-    */
-  private[speedy] def executePure(args: util.ArrayList[SValue]): SValue
 }
 
 private[speedy] sealed abstract class OnLedgerBuiltin(arity: Int)
     extends SBuiltin(arity)
     with Product {
 
-  protected def execute(
+  /** On ledger builtins may reference the Speedy machine's ledger state.
+    *
+    * @param args arguments for executing the builtin
+    * @param machine the Speedy machine (machine state may be modified by the builtin)
+    * @param onLedger a reference to the Speedy machine's ledger state (which may be modified by the builtin)
+    * @return the builtin execution's resulting control value
+    */
+  protected def executeWithLedger(
```

Comment:
Renamed to better indicate the difference of usage here

Reply:
:+1: